### PR TITLE
TPM counter support on imx93 Cortex-A Core

### PR DESCRIPTION
**Pull request by JiafeiPan**
1. Cleaned up imx_ccm_rev2 header file.
2. added TPM clock support in imx_ccm_rev2.
3. Added new counter driver for TPM
4. Added TPM device tree nodes on imx93 platform.
5. Enabled diver testing for tests/drivers/counter/counter_basic_api/

### CODE REVIEW

File: tests/drivers/counter/counter_basic_api/src/test_counter.c [hunk 1]
```diff
@@ -382,7 +385,7 @@ static void test_single_shot_alarm_instance(const struct device *dev, bool set_t
 	ticks = counter_us_to_ticks(dev, counter_period_us);
 	top_cfg.ticks = ticks;
 
-	cntr_alarm_cfg.flags = 0;
+	cntr_alarm_cfg.flags = COUNTER_ALARM_CFG_ABSOLUTE;
```
Comment (JiafeiPan) [inline tests/drivers/counter/counter_basic_api/src/test_counter.c:388]: With this update, it caused tests with twister-build failed:

```
INFO    - 102/651 nrf52_bsim                tests/drivers/counter/counter_basic_api/drivers.counter.basic_api.nrf_zli  FAILED Failed (native 0.195s)
...
INFO    - /__w/zephyr/zephyr/twister-out/nrf52_bsim/tests/drivers/counter/counter_basic_api/drivers.counter.basic_api.nrf_zli/handler.log
INFO    - 103/651 nrf52_bsim                tests/drivers/counter/counter_basic_api/drivers.counter.basic_api  FAILED Failed (native 0.318s)
INFO    - /__w/zephyr/zephyr/twister-out/nrf52_bsim/tests/drivers/counter/counter_basic_api/drivers.counter.basic_api/handler.log
...
```
If the update for test case is correct, maybe need the driver to check the failure issue.

**Comment by mmahadevan108** [comment]
Please help fix twister issues.

File: west.yml
Comment (mmahadevan108) [inline west.yml:1]: Please fix the commit title and body with some detail on what is included as part of this update

File: dts/arm64/nxp/nxp_mimx93_a55.dtsi [hunk 2]
```diff
@@ -346,6 +346,28 @@
 		dma-names = "tx", "rx";
 		status = "disabled";
 	};
+
```
Comment (mmahadevan108) [inline dts/arm64/nxp/nxp_mimx93_a55.dtsi:349]: Are there other instances of TPM in this SoC. Can we define them all

**Comment by JiafeiPan** [comment]
1. Addessed review comments.
2. Removed update for tests/drivers/counter/counter_basic_api, will submit in another PR.

**Comment by JiafeiPan** [comment]
> Please help fix twister issues.

Removed update tests/drivers/counter/counter_basic_api and will submit in another PR, so there will be no twister issue in this PR.

**Comment by mmahadevan108** [comment]
Please update `west.yml`

**Comment by JiafeiPan** [comment]
> Please update `west.yml`

Done, thanks.

**Comment by mmahadevan108** [comment]
How was this tested?

**Comment by JiafeiPan** [comment]
> How was this tested?

It can be tested with tests/drivers/counter/counter_basic_api, the following is testing log:

```
*** Booting Zephyr OS build v3.6.0-2300-g3b800d476e39 ***
Running TESTSUITE counter_basic
===================================================================
START - test_all_channels
Testing tpm@44320000
 PASS - test_all_channels in 0.133 seconds
===================================================================
START - test_cancelled_alarm_does_not_expire
Skipped for tpm@44320000
 PASS - test_cancelled_alarm_does_not_expire in 0.103 seconds
===================================================================
START - test_late_alarm
Skipped for tpm@44320000
 PASS - test_late_alarm in 0.103 seconds
===================================================================
START - test_late_alarm_error
Skipped for tpm@44320000
 PASS - test_late_alarm_error in 0.103 seconds
===================================================================
START - test_multiple_alarms
Skipped for tpm@44320000
 PASS - test_multiple_alarms in 0.103 seconds
===================================================================
START - test_set_top_value_with_alarm
Testing tpm@44320000
 PASS - test_set_top_value_with_alarm in 0.211 seconds
===================================================================
START - test_short_relative_alarm
Skipped for tpm@44320000
 PASS - test_short_relative_alarm in 0.104 seconds
===================================================================
START - test_single_shot_alarm_notop
Testing tpm@44320000
 PASS - test_single_shot_alarm_notop in 0.172 seconds
===================================================================
START - test_single_shot_alarm_top
Testing tpm@44320000

    Assertion failed at ZEPHYR_BASE/tests/drivers/counter/counter_basic_api/src/test_counter.c:419: test_single_shot_alarm_instance: (0 not equal to err)
tpm@44320000: Counter set alarm failed (err: -22)
 FAIL - test_single_shot_alarm_top in 0.020 seconds
===================================================================
TESTSUITE counter_basic failed.
Running TESTSUITE counter_no_callback
===================================================================
START - test_set_top_value_without_alarm
Testing tpm@44320000
 PASS - test_set_top_value_without_alarm in 0.108 seconds
===================================================================
TESTSUITE counter_no_callback succeeded

------ TESTSUITE SUMMARY START ------

SUITE FAIL -  88.89% [counter_basic]: pass = 8, fail = 1, skip = 0, total = 9 duration = 1.052 seconds
 - PASS - [counter_basic.test_all_channels] duration = 0.133 seconds
 - PASS - [counter_basic.test_cancelled_alarm_does_not_expire] duration = 0.103 seconds
 - PASS - [counter_basic.test_late_alarm] duration = 0.103 seconds
 - PASS - [counter_basic.test_late_alarm_error] duration = 0.103 seconds
 - PASS - [counter_basic.test_multiple_alarms] duration = 0.103 seconds
 - PASS - [counter_basic.test_set_top_value_with_alarm] duration = 0.211 seconds
 - PASS - [counter_basic.test_short_relative_alarm] duration = 0.104 seconds
 - PASS - [counter_basic.test_single_shot_alarm_notop] duration = 0.172 seconds
 - FAIL - [counter_basic.test_single_shot_alarm_top] duration = 0.020 seconds

SUITE PASS - 100.00% [counter_no_callback]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.108 seconds
 - PASS - [counter_no_callback.test_set_top_value_without_alarm] duration = 0.108 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION FAILED
```
One of the test case fails, I found the root cause is we need to update test case to use absolute alarm in this use case, but with this update it will cause the other platform's nrf52_bsim  testing fail, so we need the owner of nrf52_bsim to check the their driver to fix this issue.  I submitted these update in another PR to address these issues: https://github.com/zephyrproject-rtos/zephyr/pull/71298

**Comment by JiafeiPan** [comment]
> How was this tested?

Can be tested with tests/drivers/counter/counter_basic_api/, have enabled it for imx93_evk.

**Comment by JiafeiPan** [comment]
Rebased.

File: drivers/counter/counter_mcux_tpm.c [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+void mcux_tpm_isr(const struct device *dev)
+{
+	TPM_Type *base = get_base(dev);
+	struct mcux_tpm_data *data = dev->data;
+	uint32_t current = TPM_GetCurrentTimerCount(base);
+	uint32_t status;
+
+	status =  TPM_GetStatusFlags(base) & (kTPM_Chnl0Flag | kTPM_TimeOverflowFlag);
+	TPM_ClearStatusFlags(base, status);
+	__DSB();
```
Comment (danieldegrasse) [inline drivers/counter/counter_mcux_tpm.c:138]: I think we can use `z_barrier_dsync_fence_full` here to make this code architecture generic

**Comment by mmahadevan108** [comment]
Please help address merge conflicts

**Comment by JiafeiPan** [comment]
> Please help address merge conflicts

Rebased.

**Review by mmahadevan108 — DISMISSED** [review]

**Comment by mmahadevan108** [comment]
@danieldegrasse can you help re-review

**Review by danieldegrasse — DISMISSED** [review]

File: drivers/counter/CMakeLists.txt [hunk 4]
```diff
@@ -28,6 +28,7 @@ zephyr_library_sources_ifdef(CONFIG_COUNTER_CMOS                counter_cmos.c)
 zephyr_library_sources_ifdef(CONFIG_COUNTER_MCUX_GPT            counter_mcux_gpt.c)
```
Comment (dbaluta) [inline drivers/counter/CMakeLists.txt:28]: signed-off-by should be placed at the end.

File: drivers/counter/Kconfig.mcux_tpm [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# MCUXpresso SDK TPM
+
+# Copyright 2023 NXP
+# SPDX-License-Identifier: Apache-2.0
+
```
Comment (dbaluta) [inline drivers/counter/Kconfig.mcux_tpm:5]: 2024 now

File: drivers/counter/counter_mcux_tpm.c [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+	struct mcux_tpm_data *data = dev->data;
+	uint32_t ticks = alarm_cfg->ticks;
+
+	if (chan_id != kTPM_Chnl_0) {
+		LOG_ERR("Invalid channel id");
+		return -EINVAL;
+	}
+
+	if (ticks > (top_value)) {
+		return -EINVAL;
```
Comment (dbaluta) [inline drivers/counter/counter_mcux_tpm.c:90]: no need for `()` around top_value.

File: dts/bindings/timer/nxp,tpm-timer.yaml [hunk 7]
```diff
@@ -0,0 +1,32 @@
+# Copyright 2023 NXP
+# SPDX-License-Identifier: Apache-2.0
+
```
Comment (dbaluta) [inline dts/bindings/timer/nxp,tpm-timer.yaml:3]: 2024

**Comment by JiafeiPan** [comment]
Rebased and addressed @dbaluta's comments

**Review by danieldegrasse — APPROVED** [review]

**Review by dleach02 — APPROVED** [review]

**Comment by JiafeiPan** [comment]
@nordic-krch @mmahadevan108 could this PR been merged? thanks.

**Review by nordic-krch — APPROVED** [review]

**Pull request merged**